### PR TITLE
flow step retry feature

### PR DESCRIPTION
**Pull request by sqwishy**
When a job fails and a retry is possible, `push_next_flow_job` creates a new job scheduled for some time based on the retry configuration.

The previous job `args` are not reused because there were some weird things happening to do with for loops running using the wrong arguments passed in. So the `last_result`/`previous_result` is stored in the retry status so that it can be replayed later if a retry is necessary in order to re-calculate the job `args` again.

### CODE REVIEW

File: backend/src/worker.rs [hunk 1]
```diff
@@ -240,7 +241,7 @@ pub async fn run_worker(
                             }
                         }
                     }
-                    tracing::error!(job_id = %job.id, "Error handling job: {err}");
+                    tracing::error!(job_id = %job.id, err = err.alt(), "error handling job");
```
Comment (rubenfiszel) [inline backend/src/worker.rs:244]: The errors are also displayed in the frontend, need to check what impact this has

File: backend/src/worker_flow.rs [hunk 2]
```diff
@@ -192,14 +202,52 @@ pub async fn update_flow_status_after_job_completion(
...
+                    .and_then(|v| serde_json::from_value::<FlowValue>(v.clone()).ok())
+                    .map(|module| module.retry)
+                    .unwrap_or_default(),
+                &flow_job
+                    .flow_status
+                    .as_ref()
+                    .and_then(|v| serde_json::from_value::<FlowStatus>(v.clone()).ok())
+                    .map(|status| status.retry)
+                    .unwrap_or_default(),
+            )
```
Comment (rubenfiszel) [inline backend/src/worker_flow.rs]: looks pretty redundant, anyway to factor this ?

**Comment by sqwishy** [comment]
It should show error causes, like messages from underlying errors wrapped using `context()` into `anyhow::Error`. But this might be a bit unpredictable if we are trying to hide underlying errors like from sqlx.

It made debugging a bit easier for a bit but i wouldn't miss it if we want remove it if we dont want to spend time checking if the impact is negative

On August 27, 2022 11:56:55 PM PDT, Ruben Fiszel ***@***.***> wrote:
***@***.*** commented on this pull request.

File: backend/src/worker_flow.rs [hunk 3]
```diff
@@ -449,35 +507,113 @@ async fn push_next_flow_job(
...
+            .bind(json!(RetryStatus { fail_count, ..status.retry.clone() }))
+            .bind(flow_job.id)
+            .execute(db)
+            .await
+            .context("update flow retry")?;
+
+            /* it might be better to retry the job using the previous args instead of determining
+             * them again from the last result, but that seemed to not play well with the forloop
+             * logic and I couldn't figure out why. */
```
Comment (rubenfiszel) [inline backend/src/worker_flow.rs:561]: let's sync about this

File: backend/src/worker.rs [hunk 4]
```diff
@@ -1665,6 +1662,274 @@ def main():
...
+            }
+        }
+
+        fn flow() -> FlowValue {
+            let inner = r#"
+            export async function main(index, port) {
+                const buf = new Uint8Array([0]);
+                const sock = await Deno.connect({ port });
+                await sock.write(new Uint8Array([index]));
+                if (1 != await sock.read(buf)) throw "read";
```
Comment (rubenfiszel) [inline backend/src/worker.rs]: very nit but: `1 != await sock.read(buf)` seems harder to read than `await sock.read(buf) != 1`

File: backend/src/worker.rs [hunk 5]
```diff
@@ -1665,6 +1662,274 @@ def main():
...
+                (0, Some(99)),
+                (1, Some(99)),
+                (2, None),
+                /* pass pass pass */
+                (0, Some(3)),
+                (1, Some(5)),
+                (2, Some(7)),
+                /* fail the last step once */
+                (0xff, None),
+                (0xff, Some(9)),
```
Comment (rubenfiszel) [inline backend/src/worker.rs:1817]: i've re-read the tests a few times and I couldn't make much sense of the meaning of 99, and 3,5,7 and attempts tests

File: backend/src/worker_flow.rs [hunk 6]
```diff
@@ -302,6 +350,16 @@ async fn has_failure_module<'c>(
     .map_err(|e| Error::InternalErr(format!("error during retrieval of has_failure_module: {e}")))
 }
 
+fn next_retry(retry: &Retry, status: &RetryStatus) -> Option<(u16, Duration)> {
+    const MAX_FAILURES: u16 = 1000;
+    const HOURS: Duration = Duration::from_secs(60 * 60);
```
Comment (rubenfiszel) [inline backend/src/worker_flow.rs]: We should error if the flow is ill-defined when it is started rather than clipping acceptable values.

**Review by rubenfiszel — COMMENTED** [review]
Lots of great things here, would prefer to merge #491 first

File: backend/src/worker.rs (hunk 5 above)
Comment (sqwishy) [reply backend/src/worker.rs:1817]: The particular numbers are arbitrary, but the values from the first two tries should be rejected so having them be 99 should help us know, if we see it, that it shouldn't be there and where it came from.

The `attempts` assert is to verify the steps are running/re-running in the right order.

File: backend/src/worker.rs (hunk 1 above)
Comment (sqwishy) [reply backend/src/worker.rs:244]: `tracing::error!` shouldn't end up anywhere near the frontend/user? (or does this end up in the job logs somehow -- github won't show me context?)

**Review by rubenfiszel — COMMENTED** [review]
Missing the openapi spec changes

**Review by rubenfiszel — APPROVED** [review]

**Pull request merged**